lib/charms/kafka/v0/kafka_snap.py

```diff
@@ -0,0 +1,129 @@
+            apt.update()
+            apt.add_package("snapd")
+            cache = snap.SnapCache()
+            kafka = cache["kafka"]
+
+            if not kafka.present:
+                kafka.ensure(snap.SnapState.Latest, channel="rock/edge")
+
+            self.kafka = kafka 
+            return MaintenanceStatus("sucessfully installed kafka snap")
+
+        except (snap.SnapError, apt.PackageNotFoundError):
+            return BlockedStatus("failed to install kakfa snap")
+
+    def get_kafka_apps(self) -> List:
+        """Grabs apps from the snap property."""
+        if not self.kafka:
+            logger.warning("unable to retrive kafka snap apps: kafka snap not installed")
+            apps = []
+        else:
+            apps = self.kafka.apps
+
+        return apps
+
+
+    def start_snap_service(self, snap_service: str) -> StatusBase:
+        """Starts snap service process
+
+        Args:
+            snap_service (str): The desired service to run on the unit
+                `kafka` or `zookeeper`
+        Returns:
+            MaintenanceStatus (StatusBase): If service starts successfully
+            BlockedStatus (StatusBase): If service fails to start
+        """
+        try:
+            self.kafka.start(services=[snap_service])
+            logger.info(f"successfully started snap service: {snap_service}")
+            return ActiveStatus()
+        except snap.SnapError as e:
+            logger.error(e)
+            return BlockedStatus(f"failed starting snap service: {snap_service}")
+
+
+    @staticmethod
+    def get_properties(path: str) -> Dict[str, str]:
+        """Grabs active config lines from *.properties."""
+        with open(path, "r") as f:
+            config = f.readlines()
+
+        config_map = {}
+
+        for conf in config:
+            if conf[0] != "#" and not conf.isspace():
+                logger.debug(conf.strip())
+                config_map[conf.split("=")[0]] = conf.split("=")[1].strip()
+
+        return config_map
+
+    def get_merged_properties(self, property_label: str) -> str:
+        """Merges snap config overrides with default upstream *.properties.
+
+        Args:
+            property_label (str): The prefix of the desired *.properties file
+                e.g `server` or `zookeeper`
+
+        Returns:
+           str: The merged default and user config
+        """
+
+        default_path = self.default_config_path + f"{property_label}.properties"
+        default_config = self.get_properties(path=default_path)
+
+        override_path = self.snap_config_path + f"{property_label}.properties"
+
+        try:
+            override_config = self.get_properties(path=override_path)
+            final_config = {**default_config, **override_config}
+        except FileNotFoundError:
+            logging.info("no manual config found")
+            final_config = default_config
+
+        msg = ""
+        for k, v in final_config.items():
+            msg = f"{msg}{k}={v}\n"
```

Comment:
Is there a reason you're repeatedly assigning values to the same variable? The suggestion below is maybe less readable than your existing code, so it's just a suggestion. 
```suggestion
return "\n".join([f"{key}={value}" for key, value in final_config.items()])
```

Reply:
Really good suggestion!